streamlit_app.py

```diff
@@ -1,231 +1,248 @@
+        new_size = tf.cast(tf.convert_to_tensor(base_shape[:-1]), tf.float32) * (
+            octave_scale**octave
+        )
+        new_size = tf.cast(new_size, tf.int32)
+        img = tf.image.resize(img, new_size)
+        for step in range(steps_per_octave):
+            gradients = get_tiled_gradients(img, new_size)
+            img = img + gradients * step_size
+            img = tf.clip_by_value(img, -1, 1)
             p += 1
-            progress_widget.progress(p / (octave_n * iter_n))
-
-            write_image(image_widget, img_in)
-            text_widget.text(text_template % (octave, i))
-
-
-layers = [
-    op.name for op in graph.get_operations()
-    if op.type == 'Conv2D' and 'import/' in op.name
-    ]
-
-
-@st.cache()
-def read_file_from_url(url):
-    return urllib.request.urlopen(url).read()
-
-
-# Sidebar controls:
-
-# Temporary config option to remove deprecation warning.
-st.set_option('deprecation.showfileUploaderEncoding', False)
-
-MAX_IMG_WIDTH = 600
-MAX_IMG_HEIGHT = 400
-DEFAULT_IMAGE_URL = 'https://i.imgur.com/dOPMzXl.jpg'
-
-
-file_obj = st.sidebar.file_uploader('Choose an image:', ('jpg', 'jpeg'))
-
-if not file_obj:
-    file_obj = BytesIO(read_file_from_url(DEFAULT_IMAGE_URL))
-
-img_in = PIL.Image.open(file_obj)
-
-img_in.thumbnail((MAX_IMG_WIDTH, MAX_IMG_HEIGHT), PIL.Image.ANTIALIAS)
-img_in = np.float32(img_in)
-
-
-# Picking some internal layer. Note that we use outputs before applying the
-# ReLU nonlinearity to have non-zero gradients for features with negative
-# initial activations.
-
-max_value = len(layers) - 1
-layer_num = st.sidebar.slider('Layer to visualize', 0, max_value, min(58, max_value))
-layer = layers[layer_num]
-
-channels = int(get_tensor(layer).get_shape()[-1])
-max_value = channels - 1
-channel = st.sidebar.slider('Channel to visualize', 0, max_value, min(62, max_value))
-
-octaves = st.sidebar.slider('Octaves', 1, 30, 4)
-
-iterations = st.sidebar.slider('Iterations per octave', 1, 30, 10)
-
-
-# Show original image and final image, computing DeepDream on it iteratively.
-
-'## Original image'
-
-write_image(st, img_in)
-
-
-'## Output'
-
-out = do_deepdream(
-    get_tensor(layer)[:, :, :, channel], img_in, octave_n=octaves,
-    iter_n=iterations)
+            progress_widget.progress(
+                p / (steps_per_octave * len(octaves)),
+                text_template % (octave, step + 1),
+            )
+            if step % 10 == 0:
+                show(image_widget, deprocess(img))
+    result = PIL.Image.fromarray(np.array(deprocess(img)))
+    return result
+
+
+# Load and cache the base model and all layers
+base_model = load_base_model()
+all_layers = load_all_layers(base_model)
+
+st.sidebar.caption(
+    """This Streamlit app is based entirely on
+[TensorFlow's DeepDream tutorial](https://github.com/tensorflow/docs/blob/9ae740ab7b5b3f9c32ca060332037b51d95674ae/site/en/tutorials/generative/deepdream.ipynb)."""
+)
+
+with st.sidebar.form(key="my_form"):
```

Comment:
Why move this stuff into a form?

Part of the fun with this app is tweaking these sliders to see what happens, and moving them into a form makes this interaction worse.

Reply:
I see what you mean. The app does nothing until user interaction with the submit button, which removes the fun. I initially added it so that the model wouldn't run until the button click, thinking this would make the app appear less sluggish. But I concur with your comment and have removed the form on 010ddb8.